pkg/graphviz/traceflow.go

```diff
@@ -15,64 +15,95 @@
-	node.SetColor(dimGrey)
-	return node
+// createDirectedEdgeWithDefaultStyle creates a node with default style (usually used to represent a component in traceflow) .
+func createNodeWithDefaultStyle(graph *gographviz.Graph, parentGraph string, name string) (*gographviz.Node, error) {
+	err := graph.AddNode(parentGraph, name, map[string]string{
+		"shape": "box",
+		"style": `"rounded,filled,solid"`,
+		"color": dimGrey,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return graph.Nodes.Lookup[name], nil
 }
 
+// createEndpointNodeWithDefaultStyle creates an endpoint node with default style.
 // EndpointNode is the the type of node for endpoints in networks like the sender and the receiver.
-func createEndpointNodeWithDefaultStyle(graph *cgraph.Graph, name string) *cgraph.Node {
-	node, _ := graph.CreateNode(name)
-	node.SetColor(grey)
-	node.SetFillColor(lightGrey)
-	node.SetStyle(cgraph.FilledNodeStyle + "," + cgraph.BoldNodeStyle)
-	return node
+func createEndpointNodeWithDefaultStyle(graph *gographviz.Graph, parentGraph string, name string) (*gographviz.Node, error) {
+	err := graph.AddNode(parentGraph, name, map[string]string{
+		"style":     `"filled,bold"`,
+		"color":     grey,
+		"fillcolor": lightGrey,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return graph.Nodes.Lookup[name], nil
 }
 
-func createEdgeWithDefaultStyle(graph *cgraph.Graph, name string, start *cgraph.Node, end *cgraph.Node) *cgraph.Edge {
-	edge, _ := graph.CreateEdge(name, start, end)
-	edge.SetPenWidth(2.0)
-	edge.SetColor(silver)
-	return edge
+// createDirectedEdgeWithDefaultStyle creates a directed edge with default style.
+// It is allowed to create duplicate edges.
+func createDirectedEdgeWithDefaultStyle(graph *gographviz.Graph, start *gographviz.Node, end *gographviz.Node, isForwardDir bool) (*gographviz.Edge, error) {
+	err := graph.AddEdge(start.Name, end.Name, true, map[string]string{
+		"penwidth": "2.0",
+		"color":    silver,
+	})
+	if err != nil {
+		return nil, err
+	}
+	edges := graph.Edges.SrcToDsts[start.Name][end.Name]
+	if len(edges) == 0 {
+		return nil, errors.New(fmt.Sprintf("Failed to create a new edge between node %s and node %s", start.Name, end.Name))
+	}
+	edge := edges[len(edges)-1]
+	if isForwardDir {
+		edge.Attrs[gographviz.Dir] = "forward"
+	} else {
+		edge.Attrs[gographviz.Dir] = "back"
+	}
+	return edge, nil
 }
 
+// createDirectedEdgeWithDefaultStyle creates a cluster with default style.
 // In Graphviz, cluster is a subgraph which is surrounded by a rectangle and the nodes belonging to the cluster are drawn together.
-func createClusterWithDefaultStyle(graph *cgraph.Graph, name string) *cgraph.Graph {
-	cluster := graph.SubGraph(name, 1)
-	cluster.SetBackgroundColor(ghostWhite)
-	cluster.SetStyle(cgraph.FilledGraphStyle + "," + cgraph.BoldGraphStyle)
-	return cluster
+// In traceflow, a cluster is usually used to represent a node.
```

Comment:
since the comments in this file use "node" to mean either a "Graphviz node" or a "k8s Node", would you mind disambiguating by using "k8s Node" consistently when appropriate?

Reply:
The comment is modified.